### ISSUE TITLE
RHEL support

### ISSUE THREAD

**Issue by pegasd** [pegasd/puppet-incron#11]
@markasammut requested RHEL support for the module.

Relevant PR: #10.

For now, it seems like we only need to change service name from `incron` to `incrond` on non-Debian-based systems.

**Comment by pegasd** [comment]
Hey, @markasammut, could you please provide the output of `sudo facter -p os` from one of your systems? Preferably the most used version of RHEL for you.

**Comment by markasammut** [comment]
# facter os
{
  architecture => "x86_64",
  family => "RedHat",
  hardware => "x86_64",
  name => "RedHat",
  release => {
    full => "7.5",
    major => "7",
    minor => "5"
  },
  selinux => {
    enabled => false
  }
}

This though should apply also on CentOS hosts, so I'd use the OS Family not the OS name.

**Comment by pegasd** [comment]
Thanks, @markasammut !

Yes, good idea about the `os.family` part.

I just spun up a CentOS docker to get facter info from there.

on CentOS:
```
{
  architecture => "x86_64",
  family => "RedHat",
  hardware => "x86_64",
  name => "CentOS",
  release => {
    full => "7.6.1810",
    major => "7",
    minor => "6"
  },
  selinux => {
    enabled => false
  }
}
```

and on RedHat (re-posting):

```
{
  architecture => "x86_64",
  family => "RedHat",
  hardware => "x86_64",
  name => "RedHat",
  release => {
    full => "7.5",
    major => "7",
    minor => "5"
  },
  selinux => {
    enabled => false
  }
}
```

**Comment by pegasd** [comment]
Here's `hiera.yaml` from [puppetlabs-ntp](https://github.com/puppetlabs/puppetlabs-ntp):

```yaml
version: 5

defaults:
  datadir: 'data'
  data_hash: 'yaml_data'

hierarchy:
  - name: 'Full Version'
    path: '%{facts.os.name}-%{facts.os.release.full}.yaml'

  - name: 'Major Version'
    path: '%{facts.os.name}-%{facts.os.release.major}.yaml'

  - name: 'Distribution Name'
    path: '%{facts.os.name}.yaml'

  - name: 'Operating System Family'
    path: '%{facts.os.family}-family.yaml'

  - name: 'common'
    path: 'common.yaml'
```

Seems like for now we only need the `Operating System Family` bit.

**Comment by markasammut** [comment]
So are we going to externalise the service_name as a variable to be modified in Hiera? Should be good for us.

**Comment by markasammut** [comment]
All we'd require is this line in init.pp list of parameters:
String                $service_name  = 'incron',

And this change in service.pp:
    service { $::incron::service_name:

**Comment by pegasd** [comment]
@markasammut The idea is to have a bit of relevant Hiera inside the module, so no external configuration is required. This is similar to `params.pp`, but a lot more readable.

**Comment by pegasd** [comment]
This bit:

```
service { $::incron::service_name:
```

is better managed through `$name` parameter:

```
service { 'incron':
  name => $::incron::service_name,
...
```

so that we can refer to the service as `Service[incron]` both in specs or in manifests.

**Comment by pegasd** [comment]
Also, just for reference:

```ruby
{
  architecture => "amd64",
  distro => {
    codename => "bionic",
    description => "Ubuntu 18.04.3 LTS",
    id => "Ubuntu",
    release => {
      full => "18.04",
      major => "18.04"
    }
  },
  family => "Debian",
  hardware => "x86_64",
  name => "Ubuntu",
  release => {
    full => "18.04",
    major => "18.04"
  },
  selinux => {
    enabled => false
  }
}
```

**Comment by markasammut** [comment]
I've never used this method, so I'm asking, just to confirm, the hiera.yaml placed within the module does not replace or overwrite the global hiera.yaml we have on our Puppetmasters, does it?

**Comment by pegasd** [comment]
It works well with your own `hiera.yaml`. Chances are, some of the modules you use from Forge are already using this pattern.

You can read more documentation on the subject [here](https://puppet.com/docs/puppet/latest/hiera_migrate.html#adding-hiera-data-to-a-module).

In a few words, here's how values are substituted into classes from the lowest to the highest priority:
* module defaults
* module data (this is what we're implementing right now)
* global data (this is what you already have on your puppet server)
* class invocation

**Comment by pegasd** [comment]
@markasammut , I've created a PR here: #12 

Would be great if you tested this on your installation.

You can add something like this to your Puppetfile:

```ruby
mod 'incron',
    git:    'https://github.com/pegasd/puppet-incron',
    branch: 'GH11_redhat_support'
```

**Comment by markasammut** [comment]
Works like a charm! Well done and thanks!

**Comment by pegasd** [comment]
That's great to hear!

It'll take me some time to clean this all up and add acceptance tests for new distros.

For now, if you want to use this version, I suggest pinning a specific commit hash, like this:

```ruby
mod 'incron',
    git: 'https://github.com/pegasd/puppet-incron',
    ref: 'c9cb224'
```

Once I get acceptance tests up and running, I'll release a new version.

**Comment by pegasd** [comment]
@markasammut I'm having issues testing on CentOS.

It seems like on most other systems, both user and group creation is managed by the package, but not on CentOS. Did you have to do any extra steps on your RedHat system to get things to work?

Like doing this, for example:

```puppet
user { 'incron':
  ensure => present,
  system => true,
}

group { 'incron':
  ensure => present,
  system => true,
}
```

**Comment by pegasd** [comment]
For reference: https://travis-ci.org/pegasd/puppet-incron/jobs/586126112

**Comment by pegasd** [comment]
Phew, I believe I'm done with acceptance for this.

Added `Debian 8`, `CentOS 7`, and `Fedora 30` as test targets.

Had to chop off group management for `/var/spool/incron` and all the files inside there. Hopefully this won't break anything.

I realize there are some more acceptance scenarios that I'd love to work on, but that'll be done in a different PR (I'm currently working on switching my other module to Litmus).

**Comment by markasammut** [comment]
Ah yes, I had to add the creation of the group 'incron'. Forgot to mention it. Looks like it's no longer required now, right?

**Comment by pegasd** [comment]
That's correct, you can drop that part of your code now.

The module should only require a simple `include incron` and voila!

I released `0.7.0` version with this enhancement, so now you can update your Puppetfile to this:

```ruby
mod 'pegas/incron', '0.7.0'
```

Thanks for your help.

Enjoy!
